rune/libenclave/internal/runtime/pal/pal_linux.go

```diff
@@ -16,15 +20,49 @@ const (
 	palApiVersion = 3
 )
 
+var enclavePoolType string
+var isEncCache bool
+var hasEpm bool
+
 func (pal *enclaveRuntimePal) Init(args string, logLevel string) error {
+	var enclaveinfo *v1alpha1.Enclave
+	var err error
+	var addr int64 = 0
+	var fd int = 0
+
+	hasEpm = false
+	isEncCache = false
 	api := &enclaveRuntimePalApiV1{}
 	ver := api.get_version()
 	if ver > palApiVersion {
 		return fmt.Errorf("unsupported pal api version %d", ver)
 	}
 	pal.version = ver
-
-	return api.init(args, logLevel)
+	if strings.Contains(args, "epm") {
+		hasEpm = true
+	}
+	if hasEpm == true {
+		if strings.Contains(args, "skeleton") {
+			enclavePoolType = "enclave-cache.skeleton"
+		} else if strings.Contains(args, "occlum") {
+			enclavePoolType = "enclave-cache.occlum"
+		}
```

Comment:
Yes, I considered this just like we talked before. But when implement this, there is an issue. For exemaple when starting a container, before pal.init, it needs try to get an enclave from cache pool according to mrenclave as an argument, but how can i get mrenclave from initiated context. Mrenclave should be got after enclave measurement. 
start container--> check enclave from pool according to mrenclave(hard to retrieve) --> if yes, skip enclave init, if no, run enclave init to get mrenclave.
other solution:  adding a verification: an enclave info including mrenclave in enclave pool, when using the enclave info, we can compare mrenclave in enclave info with one in user sigstruct.

Reply:
If using server-side signing scheme, shim-rune already choses an appropriate enclave runtime (and its version), so just do it like current way, and we don't need to specify mrenclave as pool id. 

If using client-side signing scheme, client signing tool can generate collaterals and put them into the enclave image or its configuration, e.g, metadata provided by enclave image developed with Intel SGX SDK, and PAL can call sgx_get_metadata() to retrieve it. Of course this will need a specific PAL to do it, e.g, implementing a common and upper PAL loader written for Intel SGX SDK and wrapping the lower Occlum PAL for enclave runtime specific implementation.

Your solution sounds good, and we can have a discussion with the details. Currently we are talking about skeleton with client-side signing solution, so it is necessary to request the EPM pooling with an appropriate enclave pool id.